src/kOS.Safe/Screen/SubBuffer.cs

```diff
@@ -13,24 +15,116 @@ public class SubBuffer
+        /// </summary>
+        /// <param name="sb">the parent screen buffer that is being resized</param>
+        /// <returns>number of lines the parent should probably be shifted down because the subbuffer is now more lines.</returns>
+        public virtual int NotifyOfParentResize(IScreenBuffer sb)
+        {
+            ColumnCount = sb.ColumnCount;
+            return ResizeBuffer();
         }
 
-        private void InitializeBuffer()
+        protected virtual int ResizeBuffer()
         {
+            // This method also is being left open for potential overriding in subclasses
+            // if we want subbuffers that don't behave this way.  The reason is that
+            // this behavior is very specific to implementing a command line editor, and
+            // the assumption that the entire subbuffer was meant to 'flow' and wrap as one long line.
+            // That assumption may be very different for other cases, if any exist in the future.
+
+            List<char[]> newBuffer = new List<char[]>();
+            
+            int newRow = 0;
+            int oldRow = 0;
+            int newCol = 0;
+            
+            // First, copy everything from old to new, and perhaps enlarge the new if the old won't fit.
+            while (oldRow < Buffer.Count)
+            {
+                int oldCol = 0;
+                while (oldCol < Buffer[oldRow].Length)
+                {
+                    if (newCol >= ColumnCount) // wrap to new line when cur line is full.
+                    {
+                        ++newRow;
+                        newCol = 0;
+                    }
+                    if (newRow >= newBuffer.Count) // grow to required size.
+                        newBuffer.Add(new char[ColumnCount]);
+                    newBuffer[newRow][newCol] = Buffer[oldRow][oldCol];
+                    ++newCol;
+                    ++oldCol;
+                }
+                ++oldRow;
+            }
+            
+            // Then maybe append more empty rows if the above wasn't enough to fill the new size:
+            while (newBuffer.Count < RowCount)
+                newBuffer.Add(new char[ColumnCount]);
+            
+            // Reset the subbuffer row size since the new size might have had to grow (i.e. shrinking the width of a 80 col to 60 col so wrap text added a row.)
+            // int scrollDiff = (newBuffer.Count - RowCount); - this logic not quite working - disabled for now.
+
+            RowCount = newBuffer.Count;
+
+            // Because Buffer is readonly, copy the data from newBuffer into it rather than just resetting the reference to newBuffer:
             Buffer.Clear();
+            for (int i = 0 ; i < newBuffer.Count ; ++i)
+                Buffer.Add(newBuffer[i]);
 
-            for (int row = 0; row < RowCount; row++)
+            // return Fixed ? 0 : scrollDiff; - this logic not quite working - disabled for now.
+            return 0;
+        }
```

Comment:
The logic of SubBuffer.ResizeBuffer() (the method above) is important to pay attention to because it's essentially how we'd do proper line wrapping for the whole ScreenBuffer.  The important thing is that here in SubBuffer we can make the safe assumption that the entire contents of SubBuffer are indeed meant to be one joined together line.  Thus we can invoke that assumption and wrap the lines properly.  That's not an assumption we can make about the ScreenBuffer itself because the information about which rows are run-on continuation rows is lost.

The fact that we can make the assumption here is why this code is so much more complex than ResizeBuffer in ScreenBuffer.  But it also means that when you resize a terminal while a commandline was halfway typed in, that your commandline gets preserved properly and wrapped correctly even when the rest of the terminal doesn't.